dataflowengineoss/src/main/scala/io/shiftleft/dataflowengineoss/passes/reachingdef/ReachingDefProblem.scala

```diff
@@ -0,0 +1,171 @@
+  val allNodes: List[nodes.StoredNode] = method.cfgNode.toList ++ List(entryNode, exitNode) ++ method.parameter
+
+  val succ: Map[nodes.StoredNode, List[nodes.StoredNode]] = initSucc(allNodes)
+  val pred: Map[nodes.StoredNode, List[nodes.StoredNode]] = initPred(allNodes, method)
+
+  /**
+    * Create a map that allows CFG successors to be retrieved for each node
+    * */
+  private def initSucc(ns: List[nodes.StoredNode]): Map[nodes.StoredNode, List[nodes.StoredNode]] = {
+    ns.map {
+      case n @ (ret: nodes.Return)              => n -> List(ret.method.methodReturn)
+      case n @ (cfgNode: nodes.CfgNode)         => n -> cfgNode.start.cfgNext.l
+      case n @ (param: nodes.MethodParameterIn) => n -> param.start.method.cfgFirst.l
+      case n =>
+        logger.warn(s"Node type ${n.getClass.getSimpleName} should not be part of the CFG");
+        n -> List()
+    }.toMap
+  }
+
+  /**
+    * Create a map that allows CFG predecessors to be retrieved for each node
+    * */
+  private def initPred(ns: List[nodes.StoredNode],
+                       method: nodes.Method): Map[nodes.StoredNode, List[nodes.StoredNode]] = {
+    ns.map {
+      case n @ (_: nodes.CfgNode) if method.start.cfgFirst.headOption().contains(n) =>
+        n -> method.parameter.l.sortBy(_.order).lastOption.toList
+      case n @ (cfgNode: nodes.CfgNode) => n -> cfgNode.start.cfgPrev.l
+      case n @ (param: nodes.MethodParameterIn) =>
+        if (param.order == 1) { n -> List(method) } else {
+          n -> method.parameter.order(param.order - 1).headOption.toList
+        }
+      case n =>
+        logger.warn(s"Node type ${n.getClass.getSimpleName} should not be part of the CFG");
+        n -> List()
+    }.toMap
+  }
+
+}
+
+class ReachingDefTransferFunction(method: nodes.Method) extends TransferFunction[Set[nodes.StoredNode]] {
+
+  private val logger: Logger = LoggerFactory.getLogger(this.getClass)
+
+  val gen: Map[nodes.StoredNode, Set[nodes.StoredNode]] = initGen(method).withDefaultValue(Set.empty[nodes.StoredNode])
+  val kill: Map[nodes.StoredNode, Set[nodes.StoredNode]] =
+    initKill(method, gen).withDefaultValue(Set.empty[nodes.StoredNode])
+
+  def initGen(method: nodes.Method): Map[nodes.StoredNode, Set[nodes.StoredNode]] = {
+    def defsMadeByCall(call: nodes.Call): Set[nodes.StoredNode] = {
+      val indicesOfDefinedOutputParams = NoResolve
+        .getCalledMethods(call)
+        .flatMap(method => method.parameter.asOutput)
+        .filter(methPO => methPO._propagateIn().hasNext)
+        .map(_.asInstanceOf[nodes.HasOrder].order.toInt)
+      call
+        ._argumentOut()
+        .asScala
+        .toList
+        .filter(node =>
+          indicesOfDefinedOutputParams.exists(_ == node.asInstanceOf[nodes.HasArgumentIndex].argumentIndex.toInt))
+        .toSet ++ {
+        if (methodForCall(call)
+              .map(method => method.methodReturn)
+              .exists(methodReturn => methodReturn._propagateIn().hasNext) ||
```

Comment:
again, nbd, but `OR`s are hard to reason about

Reply:
Yep, getting to that.